### DISCUSSION TITLE
Saving config improperly includes non-overridden values; never erases, breaks array types

### DISCUSSION THREAD

**Discussion by narration-sd** [craftcms/cms#8455] [Ideas]
### Description
In Craft 2, pressing the Save button on a plugin Settings screen writes the screen fields values into the craft_plugin settings field. In operation, thes saved values over-ride values in config/plugin-name.php files.

However, in Craft 3, the entire content of the config/plugin-name.php files gets written to the settings field -- presumably mapping in Settings screen over-rides. This has at least two visible bugs:

- array types are broken: each Save adds extra copies of array config file values into the database settings, instead of replacing them.
- removing a config file field _never_ removes it from the settings field in the database, whether the field is scalar or an array

Thus, values never disappear if ever present; and of course array values become invalid.

In addition we might ask how multi-site config from a file is handled if it is in a settings screen?

### Steps to reproduce

1. Have a coffig/plugin-name settings file
2. Have a plugin settings screen which addresses fewer fields than in file
3. Save from the plugin Settings screen, and note that the  craft_plugins settings field now has _all_ file-based config in it, not just what was on the screen.
4. Delete a field from the file, Save, and see that it remains forever present in the  settings field....the only possible remedy is to set it to null, but that doesn't disappear it...
5. Note that using an array config field in the file, as in the example below causes replication of each array field and value, every time a Save is done, and that  this propagates into any nested array values.

### Additional info -- example multiple save results below the Craft specs

PHP version	7.1.3-3+deb.sury.org~xenial+1
Database driver & version	MySQL 5.7.17-0ubuntu0.16.04.1
Image driver & version	Imagick 3.4.3RC2, ImageMagick 6.8.9-9
Craft edition & version	Craft Pro 3.0.0-beta.26
Yii version	2.0.12
Twig version	2.3.2
Guzzle version	6.2.1
Imagine version	0.7-dev

Example array file config and results after multiple Settings screen saves:

[example-with-array-multiple-Saves.txt](https://github.com/craftcms/cms/files/1290997/example-with-array-multiple-Saves.txt)

**Comment by phoob** [comment]
I've noticed the same behavior in my plugin.

**Comment by phoob** [comment]
To solve for now it I created hidden fields with null as values in the settings.twig for the fields I don't want in the database.
